src/v/storage/log_manager.cc

```diff
@@ -88,18 +88,44 @@ ss::future<> log_manager::stop() {
     return _open_gate.close()
       .then([this] {
           return ss::parallel_for_each(_logs, [](logs_type::value_type& entry) {
-              return entry.second.handle.close();
+              return entry.second->handle.close();
           });
       })
       .then([this] { return _batch_cache.stop(); });
 }
 
-static inline logs_type::iterator find_next_non_compacted_log(logs_type& logs) {
+/**
+ * `housekeeping_scan` scans over every current log in a single pass.
+ */
+ss::future<>
+log_manager::housekeeping_scan(model::timestamp collection_threshold) {
     using bflags = log_housekeeping_meta::bitflags;
-    return std::find_if(
-      logs.begin(), logs.end(), [](const logs_type::value_type& l) {
-          return bflags::none == (l.second.flags & bflags::compacted);
-      });
+
+    if (_logs_list.empty()) {
+        co_return;
+    }
+
+    while ((_logs_list.front().flags & bflags::compacted) == bflags::none) {
+        auto& current_log = _logs_list.front();
+        current_log.flags |= bflags::compacted;
+        current_log.last_compaction = ss::lowres_clock::now();
+        co_await current_log.handle.compact(compaction_config(
+          collection_threshold,
+          _config.retention_bytes(),
+          _config.compaction_priority,
+          _abort_source));
+
+        if (_logs_list.empty()) {
+            co_return;
+        }
+
+        // Make sure the current log hasn't been removed
+        // from the _logs_list.
+        if (&current_log == &_logs_list.front()) {
```

Comment:
with very few exceptions we should really try hard to avoid low-level things stuff like comparing memory raw memory addresses to do high-level operations like test for container membership.

can we avoid this by moving the choice to the back of list before we call co_await? then it is is in the correct position, and if any other operation in the system removed the log it would be responsible for removing the entry from both indicies (which i assume is already the case)

Reply:
It shouldn't be an issue moving the list repositioning to before the `co_await`. Then the memory address won't be needed. Good call.